core/editor/src/tools/ellipse.rs

```diff
@@ -58,37 +62,112 @@ impl Fsm for EllipseToolFsmState {
 			}
 
 			(EllipseToolFsmState::LmbDown, Event::MouseMove(mouse_state)) => {
+				data.drag_current = *mouse_state;
+
 				operations.push(Operation::ClearWorkingFolder);
-				operations.push(Operation::AddCircle {
-					path: vec![],
-					insert_index: -1,
-					cx: data.drag_start.x as f64,
-					cy: data.drag_start.y as f64,
-					r: data.drag_start.distance(&mouse_state),
-					style: style::PathStyle::new(None, Some(style::Fill::new(tool_data.primary_color))),
-				});
+				operations.push(make_operation(data, tool_data));
 
 				EllipseToolFsmState::LmbDown
 			}
 
 			(EllipseToolFsmState::LmbDown, Event::LmbUp(mouse_state)) => {
-				let r = data.drag_start.distance(&mouse_state.position);
-				log::info!("draw ellipse with radius: {:.2}", r);
+				data.drag_current = mouse_state.position;
+
 				operations.push(Operation::ClearWorkingFolder);
-				operations.push(Operation::AddCircle {
-					path: vec![],
-					insert_index: -1,
-					cx: data.drag_start.x as f64,
-					cy: data.drag_start.y as f64,
-					r: data.drag_start.distance(&mouse_state.position),
-					style: style::PathStyle::new(None, Some(style::Fill::new(tool_data.primary_color))),
-				});
+				operations.push(make_operation(data, tool_data));
 				operations.push(Operation::CommitTransaction);
 
 				EllipseToolFsmState::Ready
 			}
 
+			(state, Event::KeyDown(Key::KeyShift)) => {
+				data.constrain_to_circle = true;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
+			(state, Event::KeyUp(Key::KeyShift)) => {
+				data.constrain_to_circle = false;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
+			(state, Event::KeyDown(Key::KeyAlt)) => {
+				data.center_around_cursor = true;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
+			(state, Event::KeyUp(Key::KeyAlt)) => {
+				data.center_around_cursor = false;
+
+				if state == EllipseToolFsmState::LmbDown {
+					operations.push(Operation::ClearWorkingFolder);
+					operations.push(make_operation(data, tool_data));
+				}
+
+				self
+			}
+
 			_ => self,
 		}
 	}
 }
+
+fn make_operation(data: &EllipseToolData, tool_data: &DocumentToolData) -> Operation {
+	let x0 = data.drag_start.x as f64;
+	let y0 = data.drag_start.y as f64;
+	let x1 = data.drag_current.x as f64;
+	let y1 = data.drag_current.y as f64;
+
+	let (cx, cy) = if data.center_around_cursor { (x0, y0) } else { ((x0 + x1) * 0.5, (y0 + y1) * 0.5) };
```

Comment:
Is there a reason this uses `* 0.5` instead of the conceptually simpler `/ 2.0`?

Reply:
Do they get optimized into the same assembly? I've always understood that multiplication is faster than division. To my eyes, they look equally readable. I wouldn't bother changing it.

Reply:
Yes they do: https://godbolt.org/z/1rso37Tqv

Reply:
It is [different assermbly](https://godbolt.org/z/1rso37Tqv) but both only a single instruction and I'm not qualified to talk about the performance difference

Reply:
Food for thought: https://stackoverflow.com/questions/4125033/floating-point-division-vs-floating-point-multiplication

Reply:
why not just use
```rust
#[repr(C)]
union AwesomeFloat {
    u: u32,
    f: f32,
}
let fl = AwesomeFloat { f: 3.141};
unsafe { fl.u += 1<<24};
unsafe { fl.u -= 1<<24};
```
while we're already at it?

Reply:
The performance of / 2.0 is most likely worse, but the assembly does not differ when compiled with optimizations.
I'd say the slight difference in debug mode is most likely completely negligible